File: src/main/java/com/telegram/codex/conversation/memory/ChatMemoryRepository.java
package com.telegram.codex.conversation.memory;

import com.telegram.codex.persistence.ChatMemoryEntity;
import com.telegram.codex.persistence.ChatMemoryJpaRepository;

import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Repository;

@Repository
public class ChatMemoryRepository {

    private static final Logger LOGGER = LoggerFactory.getLogger(ChatMemoryRepository.class);

    private final ChatMemoryJpaRepository repository;

    public ChatMemoryRepository(ChatMemoryJpaRepository repository) {
        this.repository = repository;
    }

    public Optional<ChatMemoryRecord> find(String chatId) {
        return repository.findById(chatId).map(entity -> new ChatMemoryRecord(entity.getChatId(), entity.getMemoryText(), entity.getUpdatedAt()));
    }

    public void persist(String chatId, String memoryText) {
        String normalized = memoryText == null ? "" : memoryText.trim();
        if (normalized.isEmpty()) {
            reset(chatId);
            return;
        }
        ChatMemoryEntity entity = repository.findById(chatId).orElseGet(ChatMemoryEntity::new);
        entity.setChatId(chatId);
        entity.setMemoryText(normalized);
        entity.setUpdatedAt(System.currentTimeMillis());
        repository.save(entity);
    }

    public void reset(String chatId) {
        repository.deleteById(chatId);
        LOGGER.info("Reset chat memory chat_id={}", chatId);
    }
}

File: src/conversation/chat-memory.repository.ts
import {Injectable, Logger} from "@nestjs/common";
import {InjectRepository} from "@nestjs/typeorm";
import {Repository} from "typeorm";
import {ChatMemoryEntity} from "../database/entities";

export interface ChatMemoryRecord {
    chatId: string;
    memoryText: string | null;
    updatedAt: number;
}

@Injectable()
export class ChatMemoryRepository {
    private readonly logger = new Logger(ChatMemoryRepository.name);

    public constructor(
        @InjectRepository(ChatMemoryEntity)
        private readonly repository: Repository<ChatMemoryEntity>
    ) {}

    public async find(chatId: string): Promise<ChatMemoryRecord | null> {
        const entity = await this.repository.findOneBy({chatId});
        return entity ? {chatId: entity.chatId, memoryText: entity.memoryText, updatedAt: Number(entity.updatedAt)} : null;
    }

    public async persist(chatId: string, memoryText: string | null | undefined): Promise<void> {
        const normalized = (memoryText ?? "").trim();
        if (!normalized) {
            await this.reset(chatId);
            return;
        }
        await this.repository.save({chatId, memoryText: normalized, updatedAt: Date.now()});
    }

    public async reset(chatId: string): Promise<void> {
        await this.repository.delete({chatId});
        this.logger.log(`Reset chat memory chat_id=${chatId}`);
    }
}
